finetune/adapter_v2.py

```diff
@@ -0,0 +1,269 @@
+
+Note: If you run into a CUDA error "Expected is_sm80 to be true, but got false", uncomment the line
+`torch.backends.cuda.enable_flash_sdp(False)` in the script below (see https://github.com/Lightning-AI/lit-llama/issues/101).
+"""
+import os
+import sys
+import time
+from pathlib import Path
+import shutil
+
+import lightning as L
+import numpy as np
+import torch
+import torch.nn as nn
+
+# support running without installing as a package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from generate import generate
+from lit_llama.adapter import LLaMA, LLaMAConfig
+from lit_llama.adapter_v2 import (
+    adapter_v2_mark_only_adapter_as_trainable,
+    adapter_v2_linear_with_bias_and_scale,
+    adapter_v2_state_from_state_dict
+    )
+from lit_llama.tokenizer import Tokenizer
+from scripts.prepare_alpaca import generate_prompt
+from lightning.fabric.strategies import DeepSpeedStrategy
+
+
+eval_interval = 600
+save_interval = 1000
+eval_iters = 100
+log_interval = 1
+devices = 1
+
+# Hyperparameters
+learning_rate = 9e-3
+batch_size = 64 / devices
+micro_batch_size = 4
+gradient_accumulation_steps = batch_size // micro_batch_size
+epoch_size = 50000  # train dataset size
+num_epochs = 5
+max_iters = num_epochs * epoch_size // devices
+weight_decay = 0.02
+max_seq_length = 256  # see scripts/prepare_alpaca.py
+warmup_steps = epoch_size * 2 // micro_batch_size // devices  # 2 epoch
+
+ds_config = {
+    "train_micro_batch_size_per_gpu": micro_batch_size,
+    "gradient_accumulation_steps": gradient_accumulation_steps,
+    "zero_optimization": {"stage": 2},
+}
+
+
+def main(
+    data_dir: str = "data/alpaca", 
+    pretrained_path: str = "checkpoints/lit-llama/7B/lit-llama.pth",
+    out_dir: str = "out/adapter_v2/alpaca",
+):
+
+    fabric = L.Fabric(
+        accelerator="cuda", 
+        devices=devices, 
+        strategy=(DeepSpeedStrategy(config=ds_config) if devices > 1 else "auto"), 
+        precision="bf16-true",
+    )
+    fabric.launch()
+    fabric.seed_everything(1337 + fabric.global_rank)
+
+    if fabric.global_rank == 0:
+        os.makedirs(out_dir, exist_ok=True)
+
+    train_data, val_data = load_datasets(data_dir=data_dir)
+
+    config = LLaMAConfig(block_size=max_seq_length)
+
+    if not os.path.isfile(pretrained_path):
+        raise FileNotFoundError(
+            f"Can't find the pretrained weights at {pretrained_path}."
+            " Please follow the instructions in the README to download them."
+        )
+    checkpoint = torch.load(pretrained_path)
+
+    with fabric.init_module():
+        model = LLaMA(config)
+        # strict=False because missing keys due to adapter weights not contained in state dict
+        model.load_state_dict(checkpoint, strict=False)
+
+        for module in model.modules():
+            if isinstance(module, nn.Linear):
+                adapter_v2_linear_with_bias_and_scale(module)
+
+
+    adapter_v2_mark_only_adapter_as_trainable(model)
```

Comment:
Perhaps the name could be shortened a bit by avoiding the name "adapter" twice
```suggestion
   mark_only_adapter_v2_as_trainable(model)
```

Reply:
I like it! Done!